src/components/shared/chart/Viewport.js

```diff
@@ -88,422 +48,521 @@ require('./Viewport.css');
+    //     MarkerChartCanvas
+    //   )
+    +viewportNeedsUpdate: (
+      prevProps: ChartProps,
+      nextProps: ChartProps
+    ) => boolean,
+  |},
+  +chartProps: ChartProps,
+|};
+
+type State = {|
+  containerWidth: CssPixels,
+  containerHeight: CssPixels,
+  containerLeft: CssPixels,
+  viewportTop: CssPixels,
+  viewportBottom: CssPixels,
+  viewportLeft: UnitIntervalOfProfileRange,
+  viewportRight: UnitIntervalOfProfileRange,
+  dragX: CssPixels,
+  dragY: CssPixels,
+  isDragging: boolean,
+  isShiftScrollHintVisible: boolean,
+|};
+
+require('./Viewport.css');
+
 /**
- * About the Flow typing:
- *
- * - `Props` are the props for the returned component. It means that they are the
- *   props that the user for this component will need to specify. From the
- *   generic definition `<Props: ViewportProps>`, they must include the
- *   properties consumed par this HOC.
- *
- * - The argument, which is the augmented component (the `WrappedComponent`),
- *   needs to accept both the `InjectedProps` and some supertype of `Props`.
- *   A supertype of `Props` is an object will some properties of
- *   `Props` but not all of them.
- *   To understand what this means to Flow, we need, like sometimes, to think
- *   backwards: from the props of `WrappedComponent`, take out `InjectedProps`,
- *   and make it part of `Props`.
- *
- * So `Props` will need to hold both `ViewportProps` as said earlier and the
- * props from `WrappedComponent` that aren't `Injectedprops`.
- * This is exactly what we want Flow to check: that the user of this HOC
- * properly passes all these props!
+ * This is the type signature for the higher order component. It's easier to use generics
+ * by separating out the type definition.
  */
-export default function withChartViewport<Props: ViewportProps>(
-  WrappedComponent: React.ComponentType<InjectedProps & $Supertype<Props>>
-): React.ComponentType<Props> {
-  class ChartViewport extends React.PureComponent<Props, State> {
-    shiftScrollId: number;
-    zoomRangeSelectionScheduled: boolean;
-    zoomRangeSelectionScrollDelta: number;
-    _container: HTMLElement | null;
-    _takeContainerRef = container => (this._container = container);
-
-    constructor(props: Props) {
-      super(props);
-      (this: any)._mouseWheelListener = this._mouseWheelListener.bind(this);
-      (this: any)._mouseDownListener = this._mouseDownListener.bind(this);
-      (this: any)._mouseMoveListener = this._mouseMoveListener.bind(this);
-      (this: any)._mouseUpListener = this._mouseUpListener.bind(this);
-
-      (this: any)._setSize = this._setSize.bind(this);
-      (this: any)._setSizeNextFrame = this._setSizeNextFrame.bind(this);
-
-      this.shiftScrollId = 0;
-      this.zoomRangeSelectionScheduled = false;
-      this.zoomRangeSelectionScrollDelta = 0;
-      this._container = null;
-
-      this.state = this.getDefaultState(props);
-    }
+export type WithChartViewport<
+  ChartOwnProps: Object,
```

Comment:
I also think this "dance" between the type and the function is more complex than it should be. See also my comment on marker-chart.
That said I don't want to hold back this patch because of this, this can be worked and made simpler in a future PR.

Reply:
That future PR should include a sort of a Flow types test suite with the errors that we expect our typing system prevent (using $ExpectError like in flow-typed). This would make our flow upgrades less painful, working with this test suite rather than our full code.

Reply:
Yes, I hate the complexity of this component.

Reply:
`ChartOwnProps` is the exact same generic in both situations, so I'd prefer to keep it the same name. I really don't know how to make this type less confusing as it is :-/